### PR TITLE
[inference] Support groupwise mode of gemv kernel

### PR DESCRIPTION
**Pull request by freeliuzc**
### PR types

New features
### PR changes

OPs 
### Description

1. Support wint4 gemv func with groupwise
2. Modify weight_only_linear
3. Support groupwise mode for weight_dequantize and weight_quantize in int4 format

PCard-77383

### CODE REVIEW

File: paddle/phi/kernels/funcs/weight_only_gemv.cu [hunk 1]
```diff
@@ -24,6 +24,8 @@ limitations under the License. */
 #include "paddle/phi/common/float16.h"
 #include "paddle/phi/core/kernel_registry.h"
 
+// #define _DEBUG_WEIGHT_ONLY_GEMV
```
Comment (wwbitejotunn) [inline paddle/phi/kernels/funcs/weight_only_gemv.cu]: 这个需要保留吗?

File: paddle/phi/kernels/funcs/weight_only_gemv.cu [hunk 2]
```diff
@@ -551,39 +553,32 @@ struct WeightOnlyProperties<WeightOnlyGroupWise<GS>> {
...
-                                                   const ElemType* zeros,
+  __device__ __forceinline__ WeightOnlyScaleLoader(const T* scales,
+                                                   const T* zeros,
                                                    int initial_offset,
                                                    int stride)
       : _scales(scales), _zeros(zeros), _stride(stride) {
     _scales += initial_offset;
-#ifndef WIN32
-    // linux
-    if constexpr (Zero) {
```
Comment (wwbitejotunn) [inline paddle/phi/kernels/funcs/weight_only_gemv.cu:582]: 这个没必要删掉吧?

File: paddle/phi/kernels/funcs/weight_only_gemv.cu [hunk 3]
```diff
@@ -622,7 +609,291 @@ struct WeightOnlyScaleLoader {
...
 
   __device__ __forceinline__ int offset() { return _offset; }
-};  // NOLINT
+};
+
+template <typename T, WeightOnlyQuantType QType>
+struct WeightOnlyConverter {};
+
+template <>
+struct WeightOnlyConverter<half, WeightOnlyQuantType::Int8b> {
```
Comment (wwbitejotunn) [inline paddle/phi/kernels/funcs/weight_only_gemv.cu:626]: 这里的功能和fast_cvt_4_packed_signed_i8s_to_2_half2s是不是一样的?

File: paddle/phi/kernels/funcs/weight_only_gemv.cu [hunk 4]
```diff
@@ -622,7 +609,291 @@ struct WeightOnlyScaleLoader {
...
+                                             T* weights_f16,
+                                             T* scale,
+                                             T* zero,
+                                             int NPerBlock,
+                                             int idx) {
+#pragma unroll
+    for (int p = 0; p < 16; ++p) {
+      weights_f16[p * NPerBlock + idx] =
+          weights_vec[p / 8 + (p % 8) * 2] * scale[idx];
+#ifdef _DEBUG_WEIGHT_ONLY_GEMV
```
Comment (wwbitejotunn) [inline paddle/phi/kernels/funcs/weight_only_gemv.cu]: debug信息就不保留了吧?

File: paddle/phi/kernels/funcs/weight_only_gemv.cu (hunk 1 above)
Comment (freeliuzc) [reply paddle/phi/kernels/funcs/weight_only_gemv.cu]: 可以删掉~

File: paddle/phi/kernels/funcs/weight_only_gemv.cu (hunk 4 above)
Comment (freeliuzc) [reply paddle/phi/kernels/funcs/weight_only_gemv.cu]: 好的，下个 commit 删除掉

File: paddle/phi/kernels/funcs/weight_only_gemv.cu (hunk 2 above)
Comment (freeliuzc) [reply paddle/phi/kernels/funcs/weight_only_gemv.cu:582]: 这里加上，paddle 的 pre-commit 会报错，不让提交，我再试下是否有其他规避的方法

File: paddle/phi/kernels/funcs/weight_only_gemv.cu (hunk 3 above)
Comment (freeliuzc) [reply paddle/phi/kernels/funcs/weight_only_gemv.cu:626]: 是一样的，这里为了 int4/int8 写在一起，包了一层 struct~

**Review by heavengate — DISMISSED** [review]
LGTM for `group_size` attr

**Comment by jeff41404** [comment]
Generally speaking, adding parameters to operator needs to update checkpoint information in `paddle/phi/api/yaml/op_version.yaml`, @heavengate need to pay attention to inference compatibility risks.

**Review by qili93 — DISMISSED** [review]
LGTM for unittest.skipIf

**Review by yuanlehome — DISMISSED** [review]

**Comment by freeliuzc** [comment]
> Generally speaking, adding parameters to operator needs to update checkpoint information in `paddle/phi/api/yaml/op_version.yaml`, @heavengate need to pay attention to inference compatibility risks.

已添加

**Review by jeff41404 — DISMISSED** [review]
LGTM

**Review by Ligoml — DISMISSED** [review]
LGTM for docs

**Review by yuanlehome — DISMISSED** [review]
LGTM

**Review by heavengate — APPROVED** [review]

**Review by vivienfanghuagood — APPROVED** [review]

**Review by Ligoml — APPROVED** [review]

**Review by jeff41404 — APPROVED** [review]

**Review by heavyrain-lzy — APPROVED** [review]
LGTM for YAML

**Review by qili93 — APPROVED** [review]
LGTM for unittest.skipIf

**Review by raindrops2sea — APPROVED** [review]

**Pull request merged**